rust_src/src/buffers.rs

```diff
@@ -1982,4 +2030,268 @@ pub fn byte_char_debug_check(b: LispBufferRef, charpos: isize, bytepos: isize) {
+/// buffer to be killed as the current buffer.  If any of them returns nil,
+/// the buffer is not killed.  The hook `kill-buffer-hook' is run before the
+/// buffer is actually killed.  The buffer being killed will be current
+/// while the hook is running.  Functions called by any of these hooks are
+/// supposed to not change the current buffer.
+///
+/// Any processes that have this buffer as the `process-buffer' are killed
+/// with SIGHUP.  This function calls `replace-buffer-in-windows' for
+/// cleaning up all windows currently displaying the buffer to be killed.
+#[lisp_fn(min = "0", intspec = "bKill buffer: ")]
+pub fn kill_buffer(buffer_or_name: Option<LispBufferOrName>) -> bool {
+    let mut b = match buffer_or_name {
+        Some(buf_or_name) => LispBufferRef::from(buf_or_name),
+        None => ThreadState::current_buffer_unchecked(),
+    };
+
+    // Avoid trouble for buffers that are already dead.
+    if !b.is_live() || unsafe { thread_check_current_buffer(b.as_mut()) } {
+        return false;
+    }
+
+    // Run hooks with the buffer to be killed the current buffer.
+    let count = unsafe { SPECPDL_INDEX() };
+
+    unsafe {
+        record_unwind_protect(Some(save_excursion_restore), save_excursion_save());
+        set_buffer_internal(b.as_mut());
+    }
+
+    // First run the query functions; if any query is answered no,
+    // don't kill the buffer.
+    if !run_hook_with_args_until_failure(&mut [Qkill_buffer_query_functions]) {
+        return unbind_to(count, Qnil).into();
+    }
+
+    // Query if the buffer is still modified.
+    if is_interactive() && b.filename().is_not_nil() && b.modified_since_save() {
+        // TODO: This was AUTO_STRING.
+        let format_str = new_unibyte_string!("Buffer %s modified; kill anyway? ");
+        if !yes_or_no_p(format(&mut [format_str, b.name_]).into()) {
+            return unbind_to(count, Qnil).into();
+        }
+    }
+
+    // If the hooks have killed the buffer, exit now.
+    if !b.is_live() {
+        return unbind_to(count, Qt).into();
+    }
+
+    // Then run the hooks.
+    unsafe { run_hook(Qkill_buffer_hook) };
+    unbind_to(count, Qnil);
+
+    // If the hooks have killed the buffer, exit now.
+    if !b.is_live() {
+        return true;
+    }
+
+    // We have no more questions to ask.  Verify that it is valid
+    // to kill the buffer.  This must be done after the questions
+    // since anything can happen within do_yes_or_no_p.
+
+    // Don't kill the minibuffer now current.
+    if LispObject::from(b) == LispWindowRef::from(unsafe { minibuf_window }).contents {
+        return false;
+    }
+
+    // When we kill an ordinary buffer which shares its buffer text
+    // with indirect buffer(s), we must kill indirect buffer(s) too.
+    // We do it at this stage so nothing terrible happens if they
+    // ask questions or their hooks get errors.
+    if b.base_buffer.is_null() && b.indirections > 0 {
+        for other in LispBufferIter::new() {
```

Comment:
I like the idea of this iterator. But maybe we can come up with a name for the iteration over the global buffer list instead of the Iterator automatically walking that list.

Reply:
So something like renaming the constructor to something like `all` or `global_list` or `from_global`?

Reply:
Ideally the LispBufferIter constructor could be given a list of buffers and there would be a function or constructor/factory that would give the iterator over the global list.